source/common/http/websocket/ws_handler_impl.cc

```diff
@@ -0,0 +1,145 @@
+    return;
+  }
+
+  Upstream::ClusterInfoConstSharedPtr cluster = thread_local_cluster->info();
+
+  if (!cluster->resourceManager(Upstream::ResourcePriority::Default).connections().canCreate()) {
+    cluster->stats().upstream_cx_overflow_.inc();
+    Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+    headers->addStatic(Headers::get().Status,
+                       std::to_string(enumToInt(Http::Code::ServiceUnavailable)));
+    stream_.encodeHeaders(std::move(headers), true);
+    return;
+  }
+
+  Upstream::Host::CreateConnectionData conn_info = cluster_manager_.tcpConnForCluster(cluster_name);
+
+  // path and host rewrites
+  route_entry_->finalizeRequestHeaders(request_headers_);
+  // for auto host rewrite
+  if (route_entry_->autoHostRewrite() && !conn_info.host_description_->hostname().empty()) {
+    request_headers_.Host()->value(conn_info.host_description_->hostname());
+  }
+
+  upstream_connection_ = std::move(conn_info.connection_);
+  read_callbacks_->upstreamHost(conn_info.host_description_);
+  if (!upstream_connection_) {
+    Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+    headers->addStatic(Headers::get().Status,
+                       std::to_string(enumToInt(Http::Code::ServiceUnavailable)));
+    stream_.encodeHeaders(std::move(headers), true);
+    return;
+  }
+
+  Filter::TcpProxy::commonInitializeUpstreamConnection(cluster);
+}
+
+void WsHandlerImpl::onConnectTimeout() {
+  ENVOY_CONN_LOG(debug, "connect timeout", read_callbacks_->connection());
+  read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_connect_timeout_.inc();
+  upstream_connection_.get()->close(Network::ConnectionCloseType::NoFlush);
+  Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+  headers->addStatic(Headers::get().Status, std::to_string(enumToInt(Http::Code::GatewayTimeout)));
+  stream_.encodeHeaders(std::move(headers), true);
+}
+
+void WsHandlerImpl::onUpstreamEvent(uint32_t event) {
+  if (event & Network::ConnectionEvent::RemoteClose) {
+    read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_destroy_remote_.inc();
+  }
+
+  if (event & Network::ConnectionEvent::LocalClose) {
+    read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_destroy_local_.inc();
+  }
+
+  if (event & Network::ConnectionEvent::RemoteClose) {
+    if (connect_timeout_timer_) {
+      read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_connect_fail_.inc();
+      read_callbacks_->upstreamHost()->stats().cx_connect_fail_.inc();
+      Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+      headers->addStatic(Headers::get().Status, std::to_string(enumToInt(Http::Code::BadGateway)));
+      stream_.encodeHeaders(std::move(headers), true);
+    }
+
+    read_callbacks_->connection().close(Network::ConnectionCloseType::FlushWrite);
```

Comment:
Good question. Unfortunately, I just took this code verbatim out of common/filters/tcp_proxy.cc and I assumed that whatever code was there did all the due diligence. 

The only reason I had to duplicate the code was because of the need for a different error handling compared to standard TCP semantics (close connections).